algorithms/CPlusPlus/Trees/binary-search-tree.cpp

```diff
@@ -158,83 +159,226 @@ bool isperfect(TreeNode* root){
+    root->right=bstdelete(root->right, m->data); 
+    return (root); 
 }
 
 
 void print(TreeNode* root){
-	/**
-	 * Print the tree in an inorder fashion
-	 * 
-	 * @params: `root` root/parent node of the tree
-	 * @return: void
-	 */
-	if(root != nullptr){
-		print(root->left);
-		std::cout << root->data << " ";
-		print(root->right);
-	}	
+    /**
+     * Print the tree in an inorder fashion
+     * 
+     * @params: `root` root/parent node of the tree
+     * @return: void
+     */
+    if(root != nullptr){
+        print(root->left);
+        std::cout << root->data << " ";
+        print(root->right);
+    }   
 }
 
 void free(TreeNode* root){
-	/*
-	 * Free up the memory in the heap
-	 *
-	 * @params: `root` root/parent node of the tree
-	 */
-
-	if(root != nullptr){
-		free(root->left);
-		free(root->right);
-		delete root;
-		root = nullptr;
-	}
+    /*
+     * Free up the memory in the heap
+     *
+     * @params: `root` root/parent node of the tree
+     */
+
+    if(root != nullptr){
+        free(root->left);
+        free(root->right);
+        delete root;
+        root = nullptr;
+    }
 }
 
 int main(){
-	TreeNode* root = nullptr;
-
-	Insert(root, 37);
-	Insert(root, 19);
-	Insert(root, 4);
-	Insert(root, 22);
-	Insert(root, 51);
-	Insert(root, 55);
-	Insert(root, 42);
-	Insert(root, 20);
-	Insert(root, 11);
-	Insert(root, 2);
-	
-	print(root);
-
-	TreeNode* n = find(root, 19);
-	std::cout << "\nValue of n: " << n->data << std::endl;
-
-	if(isfull(root)) { std::cout << "The binary tree is FULL" << std::endl; }
-	else 		{ std::cout << "The binary tree is not FULL" << std::endl; }
-
-	if(isperfect(root)) { std::cout << "The binary tree is PERFECT" << std::endl; }
-	else 		   { std::cout << "The binary tree is not PERFECT" << std::endl; }
-
-	/* 
-	Tree structure
-		37
-	       /  \
-	      19  51
-	     / \   / \
-	    4  22 42 55
-	   /\	/ 	
-	  2 11  20
-
-	OUTPUT:
-	  2 4 11 19 20 22 37 42 51 55
-	  Value of n: 19
-	  The binary tree is not FULL
-	  The binary tree is not PERFECT
-	*/
-	
-
-	// free the memory
-	free(root);
-	return 0;
+    TreeNode* root = nullptr;
+
+    Insert(root, 37);
+    Insert(root, 19);
+    Insert(root, 4);
+    Insert(root, 22);
+    Insert(root, 51);
+    Insert(root, 55);
+    Insert(root, 42);
+    Insert(root, 20);
+    Insert(root, 11);
+    Insert(root, 2);
+    
+    print(root);
+
+    TreeNode* n = find(root, 19);
+    std::cout << "\nValue of n: " << n->data << std::endl;
+
+    if(isfull(root)) { std::cout << "The binary tree is FULL" << std::endl; }
+    else        { std::cout << "The binary tree is not FULL" << std::endl; }
+
+    if(isperfect(root)) { std::cout << "The binary tree is PERFECT" << std::endl; }
+    else           { std::cout << "The binary tree is not PERFECT" << std::endl; }
+
+    std::cout << "Leaf Nodes present in the binary tree are : ";
+    leaf_nodes(root);
+    std::cout<< std::endl;
+    
+    n = f_max(root);
+    if(f_max(root)!=NULL)
```

Comment:
We already have `n` holding the result of `f_max(root)`, so this function call overhead can be avoided.

```suggestion
    if(n)
```